app/uk/gov/hmrc/transitmovements/controllers/ObjectStoreURIHelpers.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.transitmovements.controllers
+
+import cats.data.EitherT
+import play.api.mvc.BaseController
+import play.api.mvc.Headers
+import uk.gov.hmrc.transitmovements.config.Constants
+import uk.gov.hmrc.transitmovements.controllers.errors.PresentationError
+import uk.gov.hmrc.transitmovements.models.ObjectStoreResourceLocation
+import uk.gov.hmrc.transitmovements.models.ObjectStoreURI
+
+import scala.concurrent.Future
+
+trait ObjectStoreURIHelpers {
+  self: BaseController =>
+
+  def extractObjectStoreResourceLocationFromHeader(headers: Headers): EitherT[Future, PresentationError, ObjectStoreResourceLocation] =
+    EitherT(
+      Future.successful(
+        for {
+          headerValue                 <- getHeader(headers.get(Constants.ObjectStoreURI))
+          objectStoreResourceLocation <- getObjectStoreResourceLocation(headerValue)
+        } yield objectStoreResourceLocation
+      )
+    )
+
+  private def getHeader(objectStoreURI: Option[String]): Either[PresentationError, ObjectStoreURI] =
+    objectStoreURI.map(ObjectStoreURI.apply).toRight(PresentationError.badRequestError("Missing X-Object-Store-Uri header value"))
+
+  private def getObjectStoreResourceLocation(objectStoreURI: ObjectStoreURI): Either[PresentationError, ObjectStoreResourceLocation] =
+    objectStoreURI.asResourceLocation
+      .toRight(
+        PresentationError.badRequestError(s"X-Object-Store-Uri header value does not start with ${ObjectStoreURI.expectedOwner}/ (got ${objectStoreURI.value})")
+      )
+
+  def extractResourceLocation(objectStoreURI: ObjectStoreURI): EitherT[Future, PresentationError, ObjectStoreResourceLocation] =
```

Comment:
extractResourceLocation and getObjectStoreResourceLocation seem to be doing the same thing?

Reply:
Different error messages. I'm not too happy with the naming but I didn't come up with anything better.